src/differencer/jsonDiff.ts

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+/* eslint-disable @typescript-eslint/no-use-before-define */
+import _ from "lodash";
+import { DiffPatcher } from "jsondiffpatch";
+import { ramlToolLogger } from "../logger";
+
+/**
+ * Top level keys/property names in flattened JSON-LD
+ */
+export const JSON_LD_KEY_GRAPH = "@graph";
+export const JSON_LD_KEY_CONTEXT = "@context";
+export const JSON_LD_KEY_ID = "@id";
+
+/**
+ * External property, added by jsondiffpath to indicate that the node property is an array
+ */
+const JSON_DIFF_PATCH_ARRAY_KEY = "_t";
+const JSON_DIFF_PATCH_ARRAY_VALUE = "a";
+
+/**
+ * Difference types that can be identified by jsondiffpath
+ */
+export enum DiffType {
+  ADDED,
+  REMOVED,
+  MODIFIED,
+  TEXT_DIFF,
+  MOVED
+}
+
+/**
+ * Class to hold differences of a JSON node
+ */
+export class NodeDiff {
+  public added?: { [key: string]: any };
+  public removed?: { [key: string]: any };
+  constructor(public id: string) {
+    this.added = {};
+    this.removed = {};
+  }
+}
+
+/**
+ * Find differences between two flattened JSON-LD AMF graphs
+ * @param left Left JSON object to compare
+ * @param right Right JSON object to compare
+ *
+ * @returns Array of NodeDiff objects
+ */
+export function findJsonDiffs(left: object, right: object): Array<NodeDiff> {
+  //Verify left and right objects conform to flattened AMF graph json structure
+  validateJson(left, "left");
+  validateJson(right, "right");
+
+  const jsonDiff = new DiffPatcher({
+    // Define an object hash function
+    objectHash: function getObjectHash(obj, index) {
+      return obj["@id"];
+    },
+    arrays: {
+      // default true, detect items moved inside the array (otherwise they will be registered as remove+add)
+      detectMove: true,
+      // default false, the value of items moved is not included in deltas
+      includeValueOnMove: true //set to true so that we can log node ids that are moved
+    },
+    textDiff: {
+      // default 60, minimum string length (left and right sides) to use text diff algorithm: google-diff-match-patch
+      minLength: 6000
+    }
+  });
+  //add plugin to include ID of the JSON node in the diff
+  jsonDiff.processor.pipes.diff.before("collectChildren", addDiffId);
+  ramlToolLogger.debug(
+    `Added plugin to include node ID in the diff: ${jsonDiff.processor.pipes.diff.list()}`
+  );
+
+  const diffs = jsonDiff.diff(left, right);
+
+  return parseDiffs(diffs);
+}
+
+/**
+ * Verify object to diff/compare conform to flattened AMF graph json structure
+ * @param obj Json provided to diff/compare
+ * @param objLoc Left/Right json in the diff
+ */
+function validateJson(obj: any, objLoc: "left" | "right"): void {
```

Comment:
Since this is usual diff, the only reference is left and right objects or whatever we want to call. And yes its only for logging.

probably we can catch the error in diffProcessor and add the file name.

Reply:
Since it's not a blocker, we can keep what we currently have. Maybe we can see later if there is a better way to handle this.

Reply:
agree.. leaving it as it is for now